### PR TITLE
net: tcp: Handle retransmitted packets from peer.

### PR DESCRIPTION
**Pull request by pfalcon**
When we receive a packet with the sequence we already seen (and
processed), the most likely cause of it is that our ACK was lost,
and peer has to retransmit that packet. Then, we should just ACK
it, because otherwise peer will retransmit it again and again,
falling into exponential backoff and hosing the entire TCP
connection.

### CODE REVIEW

**Comment by pfalcon** [comment]
This is quick and dirty patch for https://jira.zephyrproject.org/browse/ZEP-2121 . Not suitable for merging without consideration.

**Comment by pfalcon** [comment]
@andyross: Can you please look into this patch and the underlying issue https://jira.zephyrproject.org/browse/ZEP-2121 . The code for:
~~~
+               /* Don't try to reorder packets.  If it doesn't match
+                * the next segment exactly, drop and wait for
+                * retransmit
+                */
~~~
was introduced by you in 3604b2c9548ca761c338aafe830785f5ae4f00ee.

File: subsys/net/ip/net_context.c [hunk 1]
```diff
@@ -816,6 +816,24 @@ NET_CONN_CB(tcp_established)
...
+printk("Received already seen pkt - ACKing\n");
+
+	set_appdata_values(pkt, IPPROTO_TCP);
+	uint16_t data_len = net_pkt_appdatalen(pkt);
+
+		uint32_t save = context->tcp->send_ack;
+		context->tcp->send_ack = sys_get_be32(NET_TCP_HDR(pkt)->seq) + data_len;
+		/* Force sending ACK. */
+		context->tcp->sent_ack = context->tcp->send_ack - 1;
+		send_ack(context, &conn->remote_addr);
```
Comment (jukkar) [inline subsys/net/ip/net_context.c]: We could add a force option into send_ack() call so we could avoid tweaking the sent_ack variable.

Also noticed that the send_ack() is inline, we should remove that although compiler is probably smart enough to do it itself.

File: subsys/net/ip/net_context.c [hunk 2]
```diff
@@ -816,6 +816,24 @@ NET_CONN_CB(tcp_established)
 				     sys_get_be32(NET_TCP_HDR(pkt)->ack));
 	}
 
+//printk("%u %u\n", sys_get_be32(NET_TCP_HDR(pkt)->seq), context->tcp->send_ack);
+
+	/* Peer sent us packet we've already seen. Apparently, our ack was lost. */
+	if (sys_get_be32(NET_TCP_HDR(pkt)->seq) < context->tcp->send_ack) {
+printk("Received already seen pkt - ACKing\n");
+
+	set_appdata_values(pkt, IPPROTO_TCP);
```
Comment (jukkar) [inline subsys/net/ip/net_context.c]: Why do we need to set the date_len, can't we just send an ACK that is using currently active ack values?

Comment (pfalcon) [inline subsys/net/ip/net_context.c]: This just parallels what's done in "normal ACK" branch: https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/ip/net_context.c#L827 . Apparently if I added it, it didn't work w/o it, but I don't remember exactly now.

Comment (pfalcon) [inline subsys/net/ip/net_context.c]: > can't we just send an ACK that is using currently active ack values?

And yep, we can't do this (if taken literally) - we need to ACK what peer asked us for, not what we currently expect from it.

Comment (jukkar) [inline subsys/net/ip/net_context.c]: > And yep, we can't do this (if taken literally) - we need to ACK what peer asked us for, not what we currently expect from it.

I thought that sending ACK to some value would automatically acknowledge all the previous values. RFC 793 in chapter 3.3.  Sequence Numbers says "The acknowledgment mechanism employed is cumulative so that an acknowledgment of sequence number X indicates that all octets up to but not including X have been received." https://tools.ietf.org/html/rfc793#section-3.3

Comment (pfalcon) [inline subsys/net/ip/net_context.c]: > I thought that sending ACK to some value would automatically acknowledge all the previous values.

Then the explanation of what happens is less clear. Were you able to reproduce the original issue (against archive.ubuntu.com or another site) and confirm that this patch makes a difference? We can try to just resend our current ack value (without trying to "rewind" it back), and see if that still works. I'll try that a bit later unless you beat on on that.

Comment (andyross) [inline subsys/net/ip/net_context.c]: So (sorry, late to the game)... I think I understand the point here.  We don't retransmit an ACK if we already sent it once, and that's wrong.

Nonetheless Jukka is exactly right: the ACK must always be for the maximum sequence number we have received (without "holes" of course).  It's legal and expected that the stack will "skip" ACK values, we don't have to ACK every single packet (sort of the whole point of TCP, really).

Seems like the notion in the fix is correct: if we get a old/retransmitted packet we should always synchronously send an ACK (the correct ACK, not a repetition of whatever sequence number we just received!) on the assumption that the original got droped.

Comment (pfalcon) [inline subsys/net/ip/net_context.c]: @andyross : Thanks for confirming all that. My biggest concern was that maybe after your commit I dug out, you made another commit to handle that case, which later was patched into regression, and I now just trying to reinvent the wheel (as you can see, I do the homework of reading git log, but tracing complete change history is hard, and there're indeed big code move-arounds, e.g. 29b7d9c64418bf76f3751d6240e94fe18daa52cf).

So, I'm confirming that resending the current ack seq works, but otherwise leaving the structure of this patch as it is.

Comment (pfalcon) [inline subsys/net/ip/net_context.c]: Ok, confirmed to work per above, refactored, retested, will push new version soon.

**Comment by pfalcon** [comment]
The patch has been reworking per the inline comments: the latest ack number is sent per RFC793, send_ack() was update to take "force" param instead of munging with cached value. This should be ready for final review (tested also).

File: subsys/net/ip/net_context.c [hunk 3]
```diff
@@ -735,15 +735,15 @@ static inline int send_syn_ack(struct net_context *context,
 }
 
 static inline int send_ack(struct net_context *context,
```
Comment (jukkar) [inline subsys/net/ip/net_context.c]: I think we could also remove the "inline" as that is not really needed here. Probably gcc will do it automatically but still. This could be a separate patch thou.

File: subsys/net/ip/net_context.c [hunk 4]
```diff
@@ -816,6 +816,18 @@ NET_CONN_CB(tcp_established)
 				     sys_get_be32(NET_TCP_HDR(pkt)->ack));
 	}
 
+	if ((s32_t)(sys_get_be32(NET_TCP_HDR(pkt)->seq) - context->tcp->send_ack) < 0) {
```
Comment (jukkar) [inline subsys/net/ip/net_context.c]: What happens with this current check if the send_ack has wrapped in which case the check will give incorrect answer.
I wonder if it would be good idea at this point to fix the sequence check issue you mentioned in https://jira.zephyrproject.org/browse/ZEP-2120
So we could have a generic function that would check if the seq is properly greater than some other seq, and then call that function here.

**Review by jukkar — CHANGES_REQUESTED** [review]
I think we need still a bit more work with this issue.

File: subsys/net/ip/net_context.c (hunk 3 above)
Comment (pfalcon) [inline subsys/net/ip/net_context.c]: Even per my, usually strict criteria of "don't make unrelated changes", un-inlining it would fit with this patch (like, we call it from more places - don't want to inline any longer). So, will update this patch.

File: subsys/net/ip/net_context.c (hunk 4 above)
Comment (pfalcon) [inline subsys/net/ip/net_context.c]: The check above is exactly coded to give correct answer even if seq numbers wrapped. But yes, ideally instead of that more or less complex check, there would be a semantically clearer macro/incline function call. Didn't have time to do that on weekend, so coded independently like above, will work on ZEP-2120 now.

**Comment by pfalcon** [comment]
The patch updated to use net_tcp_seq_cmp(). @jukkar : Let me know if I should introduce net_tcp_seq_lesser() instead (or swap args, or whatever). Otherwise, this should be ready for merge.

**Review by jukkar — DISMISSED** [review]

**Review by jukkar — APPROVED** [review]
